docs/upgrading.md

```diff
@@ -10,6 +10,31 @@ We're building on the version `v3.4.0` a forward compatibility layer to help use
 To help on the migration process, all deprecated components are being marked with `@deprecated` and deprecated behaviour will trigger a `E_USER_DEPRECATED` error.
 However, you can also find here the instructions on how to make your code compatible with both versions.
 
+### General migration strategy
+
+Update your existing software to the latest 3.4.x version using composer:
+
+```sh
+    composer require lcobucci/jwt ^3.4
+```
+
+Then run your tests and fix all deprecation notices.
+
+Note that PHPUnit tests will only fail if you have the `E_USER_DEPRECATED` error level activated (it is a good idea to run tests using `E_ALL`, btw.), and deprecation messages during the execution of data providers will not affect the test result at all, only output text to the console, because they are called before test execution starts.
+Make sure you do not see any of them before you continue.
+
+Now you can upgrade to the latest 4.x version:
+
+```sh
+    composer require lcobucci/jwt ^4.0
+```
+
+Remember that some deprecation messages from the 3.4 version may have notified you that things still are different in 4.0, so you may find you need to adapt your own code once more at this stage.
+Not every method call in the 3.4 version is 100% identical to the 4.0 counterpart.
```

Comment:
@SvenRtbg wasn't that mitigated with the backport of multiple audiences? Do you have other examples that we need to address?

I do expect that code that uses the forward compatibility layer should simply work on v4.x.

Reply:
I think I was referring to the fact that deprecated methods, or more precisely `getClaim('aud')`, changed the return value compared to 3.3. And that difference does still exist. For a deprecated method that should not be used... it's probably unwarranted to talk about an undetermined amount of methods that "may" be different.

Reply:
I opted for removing that sentence. I do not remember enough of my migration process to have proof for this statement, and I don't want to redo that work again. :)

Reply:
Is there a reason the 4.0 thru 4.2 `Token` interface has no `claims()` method defined? I'm comparing 3.4 `Token` and 4.0 `Token` side-by-side right now to see if the method signatures match (reasonably).

Reply:
`payload()` and `signature()` also seem to be missing.

Reply:
I found them in the 4.2 variant...